File: my-solutions/0155-min-stack/solution.py
class MinStack:

    def __init__(self):
        self.mainStack = deque()
        self.minStack = deque()

    def push(self, val: int) -> None:
        if len(self.mainStack) == 0:
            self.minStack.append(val)
        else:
            self.minStack.append(min(self.minStack[-1], val))
        self.mainStack.append(val)

    def pop(self) -> None:
        if len(self.mainStack) > 0:
            self.mainStack.pop()
            self.minStack.pop()


    def top(self) -> int:
        return self.mainStack[-1]

    def getMin(self) -> int:
        return self.minStack[-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: my-solutions/0155-min-stack/solution.cpp
class MinStack {
    stack<int> numStack;
    stack<int> minNumStack;
    int minNum;
public:
    MinStack() {
        minNum = INT_MAX;
    }
    
    void push(int val) {
        this->numStack.push(val);
        this->minNum = min(this->minNum, val);
        minNumStack.push(this->minNum);
    }
    
    void pop() {
        this->numStack.pop();
        this->minNumStack.pop();
        this->minNum = minNumStack.empty() ? INT_MAX : minNumStack.top();
    }
    
    int top() {
        return this->numStack.top();
    }
    
    int getMin() {
        return this->minNumStack.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
